tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1280 @@
+    /// you do not wish to set an instance limit, leave it unspecified.
+    ///
+    /// ```should_panic
+    /// use tokio::net::windows::NamedPipeOptions;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeOptions::new().max_instances(255);
+    /// # Ok(()) }
+    /// ```
+    pub fn max_instances(&mut self, instances: usize) -> &mut Self {
+        assert!(instances < 255, "cannot specify more than 254 instances");
+        self.max_instances = instances as DWORD;
+        self
+    }
+
+    /// The number of bytes to reserve for the output buffer.
+    ///
+    /// This corresponds to specifying [nOutBufferSize].
+    ///
+    /// [nOutBufferSize]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+    pub fn out_buffer_size(&mut self, buffer: u32) -> &mut Self {
+        self.out_buffer_size = buffer as DWORD;
+        self
+    }
+
+    /// The number of bytes to reserve for the input buffer.
+    ///
+    /// This corresponds to specifying [nInBufferSize].
+    ///
+    /// [nInBufferSize]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+    pub fn in_buffer_size(&mut self, buffer: u32) -> &mut Self {
+        self.in_buffer_size = buffer as DWORD;
+        self
+    }
+
+    /// Create the named pipe identified by the name provided in [new] for use
+    /// by a server.
+    ///
+    /// This function will call the [CreateNamedPipe] function and return the
+    /// result.
+    ///
+    /// [new]: NamedPipeOptions::new
+    /// [CreateNamedPipe]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+    ///
+    /// ```
+    /// use tokio::net::windows::NamedPipeOptions;
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\tokio-named-pipe-create";
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let server = NamedPipeOptions::new().create(PIPE_NAME)?;
+    /// # Ok(()) }
+    /// ```
+    pub fn create(&self, addr: impl AsRef<OsStr>) -> io::Result<NamedPipe> {
+        // Safety: We're calling create_with_security_attributes w/ a null
+        // pointer which disables it.
+        unsafe { self.create_with_security_attributes(addr, ptr::null_mut()) }
+    }
+
+    /// Create the named pipe identified by the name provided in [new] for use
+    /// by a server.
+    ///
+    /// This is the same as [create][NamedPipeOptions::create] except that it
+    /// supports providing security attributes.
+    ///
+    /// [new]: NamedPipeOptions::new
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `attrs` points to an initialized instance of
+    /// a [SECURITY_ATTRIBUTES] structure.
+    ///
+    /// [SECURITY_ATTRIBUTES]: [crate::winapi::um::minwinbase::SECURITY_ATTRIBUTES]
+    pub unsafe fn create_with_security_attributes(
+        &self,
+        addr: impl AsRef<OsStr>,
+        attrs: *mut (),
```

Comment:
I prefer something like `*mut ()` or `*mut c_void`. It's just a dummy type.

Reply:
So while I think `*mut ()` is a good option, but it would be one we'd have to stick with for all of `1.x`. We might want to consider designing this in a way so that a `SecurityAttributes` wrapper can be more cleanly introduced later by:
* Putting it behind the unstable feature flag.
* Using a different, less defaulty name like `create_with_security_attributes_raw` and using `*mut ()`.
* Actually introducing a dummy `SecurityAttributes` that is `#[repr(C)]`, and change the signature to `attrs: &SecurityAttributes` having people rely on transmuting until someone introduces methods for modifying it as needed.

To my understanding, not introducing it doesn't appear to be an option since at least one stakeholder has communicated that they need it.

Thoughts?

Reply:
It stuck out to me, as IIRC `*mut ()` is not a FFI-safe type, so technically, this is breaking the safety requirements.

Would something like this be fine?

```rust
#[repr(transparent)]
pub struct SecurityAttributes(SECURITY_ATTRIBUTES);

impl SecurityAttributes {
  #[cfg(feature = "unstable_winapi")]
  pub fn as_raw(&mut self) -> *mut SECURITY_ATTRIBUTES { &mut self.0 }
}
````

Then you can take the unstable feature for ease, but also have the stable guarantee that transmute from a compatible type is guaranteed safe. (e.g. [the `windows` crate's definition](https://microsoft.github.io/windows-docs-rs/doc/bindings/Windows/Win32/System/SystemServices/struct.SECURITY_ATTRIBUTES.html))

Reply:
There are no safety problems with using `*mut ()` here. We cast it to an `*mut SECURITY_ATTRIBUTES` before passing it through the FFI boundary.

Reply:
Sorry, I *intended* to say that it seemed like it should UB, but reading the reference and rustinomicon it *seems* like it should be fine? (It seems odd that nothing at all is said about casting between types with different `#[repr]` and sizes, but I guess the point is that it's only when it's derefed that it matters)

I still don't like using `()` for the semantic and public API reasons, though. It's very unconventional at best: at least use `c_void`!

Reply:
When it comes to raw pointers, what you can safely do with a pointer is determined by how it was created, not by any intermediate casts. The actual type of the raw pointer has no effect. I am 100% sure that it is safe to cast `*mut SECURITY_ATTRIBUTES` to `*mut ()` or `*mut c_void` and back.

This also comes up in other contexts, e.g. turning an immutable reference into a `*mut T` raw pointer is perfectly safe as long as you don't actually mutate through the raw pointer.

Reply:
Using `*mut ()` is not unconventional — it was what people used before `c_void` was added to std. But I agree on using `c_void` over `()`.

Reply:
@udoprog You talked about adding `_raw` to the name of this method. Don't forget about it.

Reply:
`_raw` added. long Github threads are hard to follow. Thanks for reminding me!